packages/berry-pnpify/sources/index.ts

```diff
@@ -0,0 +1,176 @@
+import fs from 'fs';
+import os from 'os';
+import path from 'path';
+
+function readdirNodeModules(pnp: any, dir: string, issuer: string): string[] | undefined {
+  const dirs: { [dir: string]: string[] } = {};
+  const locator = pnp.findPackageLocator(issuer);
+  if (!locator) {
+    return undefined;
+  }
+  const info = pnp.getPackageInformation(locator);
+  if (info.packageLocation + '/' !== issuer) {
+    return undefined;
+  }
+  const packages: string[] = Array.from(info.packageDependencies.keys());
+  packages.forEach((pkg: string) => {
+    dirs[''] = dirs[''] || [];
+    if (pkg[0] === '@') {
+      const [ scopeName, packageName ] = pkg.split('/');
+      dirs[scopeName] = dirs[scopeName] || [];
+      dirs[scopeName].push(packageName);
+      if (dirs[''].indexOf(scopeName) < 0) {
+        dirs[''].push(scopeName);
+      }
+    } else {
+      dirs[''].push(pkg);
+    }
+  });
+  return dirs[dir];
+};
+
+function getIssuerForNodeModulesPath(pnp: any, pathname: string) {
+  const locator = pnp.findPackageLocator(pathname);
+  if (!locator) {
+    return undefined;
+  }
+  let issuer = pnp.getPackageInformation(locator).packageLocation;
+  const remainder = pathname.substring(issuer.length);
+  const issuers = [issuer].concat(remainder.split(/\/node_modules/));
+
+  issuers[0] += '/';
+  issuer = issuers[0];
+  for (let issuerIdx = 1; issuerIdx < issuers.length; issuerIdx++) {
+    issuers[issuerIdx] = issuers[issuerIdx].substring(1);
+  }
+
+  try {
+    for (let issuerIdx = 1; issuerIdx < issuers.length; issuerIdx++) {
+      if (issuers[issuerIdx] === '') continue;
+      if (issuers[issuerIdx] === '@types') {
+        return issuer;
+      }
+      const nextIssuer = pnp.resolveToUnqualified(issuers[issuerIdx], issuer);
+      if (!nextIssuer) {
+        return undefined;
+      }
+      issuer = nextIssuer + '/';
+    }
+    return issuer;
+  } catch (e) {
+  }
+}
+
+const pnpApiRoots: { [pnpRootDir: string]: any } = {};
+const nonPnpPaths: { [pnpRootDir: string]: boolean } = {};
+
+function getPnpApiForPath(pathname: string) {
+  const basePath = pathname.substring(0, pathname.indexOf('node_modules') - 1);
```

Comment:
I'd recommend a regex like `^(/.*?/)node_modules(/|$)` (with `match[1]` being the `basePath`). This way it'll work with paths such as `/foo/bar/hello_node_modules_world`.